Add ability to hide token in cli tools page

juanvallejo opened the pull request:
Related Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1388770
Fixes https://github.com/openshift/origin/issues/11975

![cli-tools-page-with-hide-link](https://cloud.githubusercontent.com/assets/3902875/20283018/43d894a0-aa85-11e6-99fd-8e03f5a3b45d.png)


cc @spadgett @jwforres

jwforres commented:
@spadgett mind taking this one?

spadgett commented on app/views/command-line.html:
```diff
@@ -48,7 +48,7 @@ <h1 id="cli">Command Line Tools</h1>
                     <span class="pficon pficon-warning-triangle-o" aria-hidden="true"></span>
                     <strong>A token is a form of a password.</strong>
-                    Do not share your API token.
+                    Do not share your API token. <a href="#" ng-click="toggleShowSessionToken()" ng-show="showSessionToken">Click to hide the token...</a>
```
I'd probably just say "Hide token".

You don't need the `ng-show` here since it's on the parent `div`.

You only need `href=""`. (Do you mind fixing the reveal `href` above as well?)

spadgett commented:
@jwforres Opinion on a copy to clipboard link (without needing to reveal the token)? Right now I have to click show token and then manually select the command.

jwforres commented:
wouldn't be opposed to the idea, would need to see how it looks in context

On Mon, Nov 14, 2016 at 4:25 PM, Sam Padgett <notifications@github.com>
wrote:

> @jwforres <https://github.com/jwforres> Opinion on a copy to clipboard
> link (without needing to reveal the token)? Right now I have to click show
> token and then manually select the command.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin-web-console/pull/871#issuecomment-260467450>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABZk7Tuz-Jcc1khPePpZw_-dq32STlbbks5q-NHMgaJpZM4Kx0RE>
> .
>

juanvallejo commented:
@jwforres @spadgett 
![cli-tools-page-copy-token-to-clipboard](https://cloud.githubusercontent.com/assets/3902875/20311217/f32e0e84-ab1c-11e6-81eb-387d8e3c3453.png)

![cli-tools-page-copied-token-to-clipboard](https://cloud.githubusercontent.com/assets/3902875/20311230/ff426ecc-ab1c-11e6-9aa8-2820ea45bfe3.png)

spadgett commented:
@juanvallejo +1, except I would copy the full command. Something like...?

```html
<pre>oc login https://localhost:8443 --token=&lt;hidden&gt;</pre>
<a href="">Show token</a>
<span class="action-divider">|</span>
<a href="">Copy command to clipboard</a>
```

If we made that change, we'd probably always show the warning about tokens. And might want to avoid the copy to clipboard link for browsers that don't support it like iOS Safari.

juanvallejo commented:
@spadgett Sounds good, I'll update this.
Thanks for the feedback, addressed review comments on https://github.com/openshift/origin-web-console/pull/871#pullrequestreview-8488182 . I tried updating the reveal `href` to leave `ng-show` to its parent, however its sibling `span` has `ng-show` set to an opposite value

juanvallejo commented:
@spadgett 

![cli-tools-page-showtoken-copycommand](https://cloud.githubusercontent.com/assets/3902875/20326890/b9a30412-ab58-11e6-8363-a1aa11a4a92c.png)

> `Show token` is clicked

![cli-tools-page-token-shown](https://cloud.githubusercontent.com/assets/3902875/20326912/dc9495f8-ab58-11e6-8843-6da6fa60ed0a.png)

> `Copy to clipboard` is clicked

![cli-tools-page-copy-to-clipboard](https://cloud.githubusercontent.com/assets/3902875/20326936/076774bc-ab59-11e6-9920-78a12ea19d65.png)

spadgett commented on app/scripts/controllers/commandLine.js:
```diff
@@ -15,8 +15,36 @@ angular.module('openshiftConsole')
     };
+
+    $scope.copySessionToken = function($scope) {
+      var node = $('#' + $scope.copySessionTokenId);
```
I suggest we update the `copy-to-clipboard` directive to take in an element ID or create a new directive we could reuse (scripts/directives/util.js).

spadgett commented on app/views/command-line.html:
```diff
@@ -48,7 +53,7 @@ <h1 id="cli">Command Line Tools</h1>
                   <div ng-show="showSessionToken" class="alert alert-warning">
```
Might want to show to show this always since I won't see the warning if I click copy to clipboard.

spadgett commented on app/views/command-line.html:
```diff
@@ -38,7 +38,12 @@ <h1 id="cli">Command Line Tools</h1>
+                      <span ng-show="!showSessionToken">
+                        <a href="#" ng-click="toggleShowSessionToken()">Show token</a>
+                        <span class="action-divider">|</span>
+                        <a href="#" id="{{copySessionTokenId}}" ng-click="copySessionToken(this)" title="Copy command to clipboard" data-toggle="tooltip">Copy command to clipboard</a>
```
Hide this if it's iOS? There's an `IS_IOS` constant you can use in the controller to set a scope var.

spadgett replied:
How do the actions look if you move them below the `prettyprint` div?

spadgett commented on app/views/command-line.html:
```diff
@@ -48,7 +53,7 @@ <h1 id="cli">Command Line Tools</h1>
                     <span class="pficon pficon-warning-triangle-o" aria-hidden="true"></span>
                     <strong>A token is a form of a password.</strong>
-                    Do not share your API token.
+                    Do not share your API token. <a href="#" ng-click="toggleShowSessionToken()">Hide token.</a>
```
Might want to change the show link to a hide link and just have it toggle.

juanvallejo commented:
@spadgett Review comments addressed, I'm now re-using the `<copy-to-clipboard` directive

![cli-tools-page-with-new-copy-command](https://cloud.githubusercontent.com/assets/3902875/20367532/e40e52d6-ac1d-11e6-81f7-29f5f2031937.png)

spadgett commented:
LGTM, @jwforres OK with this change?

spadgett commented:
@sspeiche FYI

jwforres commented:
yep, @spadgett were you going to do the code review on this

spadgett commented:
> were you going to do the code review on this

Yeah, code looks good. Thanks @juanvallejo

spadgett commented:
[merge]

spadgett commented:
hold, spotted one problem

spadgett commented on app/views/command-line.html:
```diff
@@ -38,14 +38,22 @@ <h1 id="cli">Command Line Tools</h1>
                   <p>
                     After downloading and installing it, you can start by logging in using<span ng-if="sessionToken"> this current session token</span>:
                     <div class="code prettyprint ng-binding" ng-if="sessionToken">
```
@juanvallejo need an outer div with `ng-if="sessionToken"` surrounding both the `prettyprint` div and the action links below.

sspeiche commented:
Why would this "copy to clipboard" not be the same icon for webhooks ?

juanvallejo commented:
@spadgett Thanks, review comment addressed

@sspeiche 
> Why would this "copy to clipboard" not be the same icon for webhooks ?

Would that be the 
![clip](https://cloud.githubusercontent.com/assets/3902875/20402587/3da0b368-accb-11e6-96b3-8184bd664ed4.png) 
icon? I did not think it would fit with this page, as I have only seen it used when copying contents from an `<input` element

sspeiche commented:
It looks like something like this:

![image](https://cloud.githubusercontent.com/assets/727196/20402673/a0050b58-accb-11e6-8dba-565195eced8a.png)

spadgett commented:
@sspeiche are you proposing to use the entire input field or just the icon?

sspeiche commented:
I'm proposing that everywhere in the web console that we "copy to clipboard" it doesn't do something that looks and feels that different.

spadgett commented:
I agree consistency is good, but we're being careful not to show the token value. Maybe the copy-to-clipboard input is not wide enough to reveal it. I'm not we should count on that, though.

jwforres commented:
if we are going to do that, we should just switch all of these code blocks on this page to copy to clipboard elements instead

@juanvallejo mind getting a screenshot of what that would look like?

juanvallejo commented:
@jwforres 
> if we are going to do that, we should just switch all of these code blocks on this page to copy to clipboard elements instead

Sure, I'll put an example together

juanvallejo commented:
@jwforres @spadgett @sspeiche 
![cli-tools-page-inputs](https://cloud.githubusercontent.com/assets/3902875/20442882/76b4d544-ad98-11e6-9fa8-96b26bd10b37.png)

sspeiche commented:
Fantastic, shouldn't the input fields be wider?

jwforres commented:
Making the input fields wider would cause the token to be visible, which is
bad.  We have them short because they are often obscuring things later in
the input, like with webhooks they are obscuring the secrets in the URL

On Fri, Nov 18, 2016 at 3:25 PM, Steve Speicher <notifications@github.com>
wrote:

> Fantastic, shouldn't the input fields be wider?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin-web-console/pull/871#issuecomment-261633171>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABZk7QAPLLceF7h2dZ06iFnwQkG9m10Yks5q_gnEgaJpZM4Kx0RE>
> .
>

jwforres commented:
That said, I'm fine with this change.  @spadgett ?

On Fri, Nov 18, 2016 at 4:09 PM, Jessica Forrester <jforrest@redhat.com>
wrote:

> Making the input fields wider would cause the token to be visible, which
> is bad.  We have them short because they are often obscuring things later
> in the input, like with webhooks they are obscuring the secrets in the URL
>
> On Fri, Nov 18, 2016 at 3:25 PM, Steve Speicher <notifications@github.com>
> wrote:
>
>> Fantastic, shouldn't the input fields be wider?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/openshift/origin-web-console/pull/871#issuecomment-261633171>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/ABZk7QAPLLceF7h2dZ06iFnwQkG9m10Yks5q_gnEgaJpZM4Kx0RE>
>> .
>>
>
>

spadgett commented:
Does it make sense to copy commands with placeholder values like `<project-name>`? (Maybe?)

spadgett commented:
> Making the input fields wider would cause the token to be visible, which is bad.

We might be able to still show `token=<hidden>` in the input and copy a different value, although we'd need to add some additional logic to the `copy-to-clipboard` directive. I'd want to check how 100% width looks in other places we use it since it's a global style.

juanvallejo commented:
@spadgett I can add a `wide` option to the directive so that full width does not apply anywhere else. I'll put together an example with this and `<hidden>` values

juanvallejo commented:
@spadgett @jwforres @sspeiche 
![cli-tools-page-wide-inputs](https://cloud.githubusercontent.com/assets/3902875/20447904/5b71e01c-adaf-11e6-9053-7325e3f8bc00.png)

Clipboard contents from copying first input `oc login https://localhost:8443 --token=YGvpJBROMXDaY_uO8XZuPKRMh3TKycqGcK9IkBXnxzE`

juanvallejo commented:
@spadgett added commit with `copy-to-clipboard` changes, PTAL

spadgett commented on app/views/directives/_copy-to-clipboard.html:
```diff
@@ -1,11 +1,17 @@
-<div class="input-group copy-to-clipboard">
-  <input id="{{id}}" type="text" class="form-control" value="{{clipboardText}}" ng-disabled="isDisabled" ng-readonly="!isDisabled" select-on-focus>
+<div class="input-group copy-to-clipboard" style="{{displayWide ? 'max-width:100%;' : ''}}">
```
Suggest something like

```
ng-class="{ 'limit-width': !displayWide }"
```

then update the CSS to key off that class.

https://github.com/spadgett/origin-web-console/blob/e01508561bb7d0c7f6c53dbedcf7dd42d37e0c52/app/styles/_core.less#L1208-L1210

spadgett commented on app/views/directives/_copy-to-clipboard.html:
```diff
@@ -1,11 +1,17 @@
+  <input id="{{id}}" type="text" class="form-control" value="{{inputText ? inputText : clipboardText}}" ng-disabled="isDisabled" ng-readonly="!isDisabled" select-on-focus>
   <span class="input-group-btn" ng-hide="hidden">
-    <a data-clipboard-target="#{{id}}"
+    <a ng-show="!inputText" data-clipboard-target="#{{id}}"
```
If you use `ng-if` here and line 10, does it let you remove the `get(0)` / `get(1)` special case in the controller?

spadgett replied:
Ah, OK. We could wrap the controller code in a `$timeout` to let the DOM render first, but what you have is probably better.

juanvallejo commented:
@spadgett thanks for the feedback, review comments addressed

spadgett commented:
thanks @juanvallejo 

LGTM, please squash, and I'll merge

juanvallejo commented:
@spadgett 
> LGTM, please squash, and I'll merge

Thanks, done!

spadgett commented:
[merge]
